From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Janmm14 <gitconfig1@janmm14.de>
Date: Tue, 11 Jan 2022 16:21:16 +0200
Subject: [PATCH] More DoS mitigations

Check the first 4 packet lengths already in Varint21FrameDecoder, Handshake
packet got minimum and maximum limits as well.

Updated to 1.19 by MrIvanPlays

Co-authored-by: Ivan Pekov <ivan@mrivanplays.com>

diff --git a/protocol/src/main/java/net/md_5/bungee/protocol/Varint21FrameDecoder.java b/protocol/src/main/java/net/md_5/bungee/protocol/Varint21FrameDecoder.java
index f33df1f37bda8327f12a55329b65f67d783c653c..14099bf062467e010d497be2052c3ed9e59ef2f1 100644
--- a/protocol/src/main/java/net/md_5/bungee/protocol/Varint21FrameDecoder.java
+++ b/protocol/src/main/java/net/md_5/bungee/protocol/Varint21FrameDecoder.java
@@ -7,9 +7,17 @@ import io.netty.handler.codec.ByteToMessageDecoder;
 import io.netty.handler.codec.CorruptedFrameException;
 import java.util.List;
 
+@lombok.RequiredArgsConstructor // IvanCord
 public class Varint21FrameDecoder extends ByteToMessageDecoder
 {
 
+    // IvanCord start
+    private final boolean limited;
+    private int packetCount = 0;
+    @lombok.Setter private int protocolVersion = -1;
+    public Varint21FrameDecoder() { this( false ); }
+    // IvanCord end
+
     private static boolean DIRECT_WARNING;
 
     @Override
@@ -46,6 +54,44 @@ public class Varint21FrameDecoder extends ByteToMessageDecoder
                 {
                     throw new CorruptedFrameException( "Empty Packet!" );
                 }
+                // IvanCord - length sanity checks for first packets in client connection if limited
+                if ( limited && packetCount < 4 )
+                {
+                    int maxLength = 2097151; // max length of 21-bit varint
+                    switch ( packetCount )
+                    {
+                        case 0:
+                            maxLength = net.md_5.bungee.protocol.packet.Handshake.EXPECTED_MAX_LENGTH + 2;
+                            break;
+                        case 1:
+                            // in case of server list ping, the packets we get after handshake are always smaller
+                            // than any of these, so no need for special casing
+                            if ( protocolVersion >= ProtocolConstants.MINECRAFT_1_19 )
+                            {
+                                maxLength = 1 + (16 * 4) + 1024;
+                                break;
+                            }
+                            maxLength = 66;
+                            break;
+                        case 2:
+                            // if offline mode we get minecraft:brand (bigger), otherwise we get EncryptionResponse
+                            // so we check for the bigger packet, we are still far below critical maximum sizes
+                            // minecraft:brand (16 bytes) followed by a 400 char long string should never be reached
+                            maxLength = 1619;
+                            break;
+                        case 3:
+                            // in offline mode we get either teleport confirm or player pos&look
+                            // otherwise we get minecraft:brand (bigger max size)
+                            maxLength = 1619;
+                            break;
+                    }
+                    if ( maxLength != -1 && length > maxLength )
+                    {
+                        com.mrivanplays.ivancord.shared.Errors.oversizedPacket( packetCount, maxLength, length );
+                    }
+                    packetCount++;
+                }
+                // IvanCord end
 
                 if ( in.readableBytes() < length )
                 {
diff --git a/protocol/src/main/java/net/md_5/bungee/protocol/packet/Handshake.java b/protocol/src/main/java/net/md_5/bungee/protocol/packet/Handshake.java
index 7dbbfd3cd0212154ce27b1134e64b715d42e8096..1e5370117770a6daf7c22de2773c6ddd98f58ec7 100644
--- a/protocol/src/main/java/net/md_5/bungee/protocol/packet/Handshake.java
+++ b/protocol/src/main/java/net/md_5/bungee/protocol/packet/Handshake.java
@@ -43,4 +43,20 @@ public class Handshake extends DefinedPacket
     {
         handler.handle( this );
     }
+
+    // IvanCord start
+    public static int EXPECTED_MAX_LENGTH = 1292;
+
+    @Override
+    public int expectedMaxLength(ByteBuf buf, net.md_5.bungee.protocol.ProtocolConstants.Direction direction, int protocolVersion)
+    {
+        return EXPECTED_MAX_LENGTH;
+    }
+
+    @Override
+    public int expectedMinLength(ByteBuf buf, net.md_5.bungee.protocol.ProtocolConstants.Direction direction, int protocolVersion)
+    {
+        return 5;
+    }
+    // IvanCord end
 }
diff --git a/protocol/src/main/java/net/md_5/bungee/protocol/packet/LoginRequest.java b/protocol/src/main/java/net/md_5/bungee/protocol/packet/LoginRequest.java
index d0e37793d1154ea08e9af1a6b8ec13942b89a918..31a2f0377c4d6ad437098fe305cb40e66dcd70c9 100644
--- a/protocol/src/main/java/net/md_5/bungee/protocol/packet/LoginRequest.java
+++ b/protocol/src/main/java/net/md_5/bungee/protocol/packet/LoginRequest.java
@@ -67,10 +67,11 @@ public class LoginRequest extends DefinedPacket
     }
 
     // Waterfall start: Additional DoS mitigations, courtesy of Velocity
+    @Override // IvanCord
     public int expectedMaxLength(ByteBuf buf, ProtocolConstants.Direction direction, int protocolVersion) {
         // Accommodate the rare (but likely malicious) use of UTF-8 usernames, since it is technically
         // legal on the protocol level.
-        if (protocolVersion >= ProtocolConstants.MINECRAFT_1_19) return -1;
+        if (protocolVersion >= ProtocolConstants.MINECRAFT_1_19) return 1 + (16 * 3) + 1024; // IvanCord
         return 1 + (16 * 3);
     }
     // Waterfall end
diff --git a/proxy/src/main/java/net/md_5/bungee/netty/ChannelWrapper.java b/proxy/src/main/java/net/md_5/bungee/netty/ChannelWrapper.java
index c6d36cdbae92f3c483d390a6e971fff301f50a8a..487c1db411e8f9ff1feb20f4ced03ecdef7c53ec 100644
--- a/proxy/src/main/java/net/md_5/bungee/netty/ChannelWrapper.java
+++ b/proxy/src/main/java/net/md_5/bungee/netty/ChannelWrapper.java
@@ -43,6 +43,7 @@ public class ChannelWrapper
 
     public void setVersion(int protocol)
     {
+        ch.pipeline().get( net.md_5.bungee.protocol.Varint21FrameDecoder.class ).setProtocolVersion( protocol ); // IvanCord
         ch.pipeline().get( MinecraftDecoder.class ).setProtocolVersion( protocol );
         ch.pipeline().get( MinecraftEncoder.class ).setProtocolVersion( protocol );
     }
diff --git a/proxy/src/main/java/net/md_5/bungee/netty/PipelineUtils.java b/proxy/src/main/java/net/md_5/bungee/netty/PipelineUtils.java
index d4f27a5dba93817753beae73b5b0be0b87b7eef2..47ea350a4904ba8816e77a0b01f806bdaa024ecf 100644
--- a/proxy/src/main/java/net/md_5/bungee/netty/PipelineUtils.java
+++ b/proxy/src/main/java/net/md_5/bungee/netty/PipelineUtils.java
@@ -79,7 +79,7 @@ public class PipelineUtils
 
 
             try {
-            BASE.initChannel( ch );
+                LIMITED_BASE.initChannel( ch ); // IvanCord - Use LIMITED_BASE
             } catch (Exception e) {
                 e.printStackTrace();
                 ch.close();
@@ -118,6 +118,7 @@ public class PipelineUtils
         // IvanCord end
     };
     public static final Base BASE = new Base();
+    public static final Base LIMITED_BASE = new Base( true ); // IvanCord - Add LIMITED_BASE for connections from clients
     private static final KickStringWriter legacyKicker = new KickStringWriter();
     private static final Varint21LengthFieldPrepender framePrepender = new Varint21LengthFieldPrepender();
     public static final String TIMEOUT_HANDLER = "timeout";
@@ -237,9 +238,18 @@ public class PipelineUtils
     private static final int HIGH_MARK = Integer.getInteger( "net.md_5.bungee.high_mark", 2 << 20 ); // 2 mb
     private static final WriteBufferWaterMark MARK = new WriteBufferWaterMark( LOW_MARK, HIGH_MARK );
 
+    @lombok.RequiredArgsConstructor // IvanCord
     public static final class Base extends ChannelInitializer<Channel>
     {
 
+        // IvanCord start
+        private final boolean fromClient;
+        public Base()
+        {
+            this( false );
+        }
+        // IvanCord end
+
         @Override
         public void initChannel(Channel ch) throws Exception
         {
@@ -255,7 +265,7 @@ public class PipelineUtils
             ch.config().setAllocator( PooledByteBufAllocator.DEFAULT );
             ch.config().setWriteBufferWaterMark( MARK );
 
-            ch.pipeline().addLast( FRAME_DECODER, new Varint21FrameDecoder() );
+            ch.pipeline().addLast( FRAME_DECODER, fromClient ? new Varint21FrameDecoder( true ) : new Varint21FrameDecoder() ); // IvanCord - limit decoding length for client packets
             ch.pipeline().addLast( TIMEOUT_HANDLER, new ReadTimeoutHandler( BungeeCord.getInstance().config.getTimeout(), TimeUnit.MILLISECONDS ) );
             ch.pipeline().addLast( FRAME_PREPENDER, framePrepender );
 
diff --git a/shared/src/main/java/com/mrivanplays/ivancord/shared/Errors.java b/shared/src/main/java/com/mrivanplays/ivancord/shared/Errors.java
index 15d4214c05eb4b83aa17665c909560d4c318f29c..77b3a0a648214663db9c6d2b688e540baf9ad5f5 100644
--- a/shared/src/main/java/com/mrivanplays/ivancord/shared/Errors.java
+++ b/shared/src/main/java/com/mrivanplays/ivancord/shared/Errors.java
@@ -15,6 +15,17 @@ public class Errors
     private final QuietException UNEXPECTED_BUFFER_END = newError( "Buffer end reached unexpectedly" );
     private final QuietException BAD_FRAME_LENGTH = newError( "length wider than 21-bit" );
     private final QuietException DISCARD = newError( "Discard handler is added to the pipeline" );
+    private final QuietException OVERSIZED_PACKET = newError( "A packet could not be framed because it was too large" );
+
+    public void oversizedPacket(int packetCount, int expected, int actual)
+    {
+        if ( DEBUG )
+        {
+            throw new CorruptedFrameException( "Packet #" + packetCount + " could not be framed because it was too large (expected "
+                + expected + " bytes, got " + actual + " bytes." );
+        }
+        throw OVERSIZED_PACKET;
+    }
 
     public void invalidProtocol(int protocol)
     {
